File: Android/app/src/main/java/com/simplemagnify/components/SettingsToggle.kt
package com.simplemagnify.components

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.semantics.stateDescription
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.simplemagnify.models.AppSettings
import com.simplemagnify.models.ButtonPosition
import com.simplemagnify.utils.Constants

@Composable
fun SettingsToggle(
    title: String,
    isOn: Boolean,
    settings: AppSettings,
    onToggle: (Boolean) -> Unit
) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(Constants.Dimensions.cornerRadius))
            .background(settings.buttonBackgroundColor)
            .clickable { onToggle(!isOn) }
            .padding(horizontal = 20.dp, vertical = 16.dp)
            .semantics {
                contentDescription = title
                stateDescription = if (isOn) "On" else "Off"
            },
        horizontalArrangement = Arrangement.SpaceBetween,
        verticalAlignment = Alignment.CenterVertically
    ) {
        Text(
            text = title,
            fontSize = Constants.FontSizes.label,
            color = settings.textColor
        )

        BigToggle(
            isOn = isOn,
            settings = settings,
            onToggle = onToggle
        )
    }
}

@Composable
fun BigToggle(
    isOn: Boolean,
    settings: AppSettings,
    onToggle: (Boolean) -> Unit
) {
    Box(
        modifier = Modifier
            .width(70.dp)
            .height(40.dp)
            .clip(RoundedCornerShape(20.dp))
            .background(if (isOn) settings.buttonColor else Constants.Colors.sliderTrack)
            .clickable { onToggle(!isOn) },
        contentAlignment = if (isOn) Alignment.CenterEnd else Alignment.CenterStart
    ) {
        Box(
            modifier = Modifier
                .padding(4.dp)
                .size(32.dp)
                .clip(CircleShape)
                .background(Color.White)
        )
    }
}

@Composable
fun SettingsSegmentedControl(
    title: String,
    selection: ButtonPosition,
    settings: AppSettings,
    onSelectionChange: (ButtonPosition) -> Unit
) {
    Column(modifier = Modifier.fillMaxWidth()) {
        Text(
            text = title,
            fontSize = Constants.FontSizes.label,
            color = settings.textColor
        )

        Spacer(modifier = Modifier.height(12.dp))

        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.spacedBy(Constants.Dimensions.buttonSpacing)
        ) {
            ButtonPosition.values().forEach { position ->
                val isSelected = selection == position
                val backgroundColor = if (isSelected) settings.buttonColor else settings.buttonBackgroundColor
                val textColor = if (isSelected) settings.backgroundColor else settings.textColor

                Box(
                    modifier = Modifier
                        .weight(1f)
                        .height(60.dp)
                        .clip(RoundedCornerShape(Constants.Dimensions.cornerRadius))
                        .background(backgroundColor)
                        .clickable { onSelectionChange(position) }
                        .semantics {
                            contentDescription = "${position.name} hand mode"
                            stateDescription = if (isSelected) "Selected" else "Not selected"
                        },
                    contentAlignment = Alignment.Center
                ) {
                    Text(
                        text = if (position == ButtonPosition.LEFT) Constants.Strings.LEFT else Constants.Strings.RIGHT,
                        fontSize = Constants.FontSizes.button,
                        fontWeight = FontWeight.Medium,
                        color = textColor
                    )
                }
            }
        }
    }
}

File: iOS/SimpleMagnify/Components/SettingsToggle.swift
import SwiftUI

struct SettingsToggle: View {
    let title: String
    @Binding var isOn: Bool

    @EnvironmentObject var settings: AppSettings

    var body: some View {
        HStack {
            Text(title)
                .font(Constants.Fonts.label)
                .foregroundColor(settings.textColor)

            Spacer()

            Toggle("", isOn: $isOn)
                .labelsHidden()
                .toggleStyle(BigToggleStyle(settings: settings))
        }
        .padding(.vertical, 16)
        .padding(.horizontal, 20)
        .background(settings.buttonBackgroundColor)
        .cornerRadius(Constants.Dimensions.cornerRadius)
        .accessibilityElement(children: .combine)
        .accessibilityLabel(title)
        .accessibilityValue(isOn ? "On" : "Off")
        .accessibilityHint("Double tap to toggle")
    }
}

struct BigToggleStyle: ToggleStyle {
    let settings: AppSettings

    func makeBody(configuration: Configuration) -> some View {
        Button(action: {
            let impactFeedback = UIImpactFeedbackGenerator(style: .light)
            impactFeedback.impactOccurred()
            configuration.isOn.toggle()
        }) {
            RoundedRectangle(cornerRadius: 20)
                .fill(configuration.isOn ? settings.buttonColor : Constants.Colors.sliderTrack)
                .frame(width: 70, height: 40)
                .overlay(
                    Circle()
                        .fill(Color.white)
                        .frame(width: 32, height: 32)
                        .shadow(color: .black.opacity(0.15), radius: 2, x: 0, y: 1)
                        .offset(x: configuration.isOn ? 14 : -14)
                        .animation(.easeInOut(duration: 0.2), value: configuration.isOn)
                )
        }
    }
}

struct SettingsSegmentedControl: View {
    let title: String
    @Binding var selection: ButtonPosition

    @EnvironmentObject var settings: AppSettings

    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            Text(title)
                .font(Constants.Fonts.label)
                .foregroundColor(settings.textColor)

            HStack(spacing: Constants.Dimensions.buttonSpacing) {
                ForEach(ButtonPosition.allCases, id: \.self) { position in
                    Button(action: {
                        let impactFeedback = UIImpactFeedbackGenerator(style: .light)
                        impactFeedback.impactOccurred()
                        selection = position
                    }) {
                        Text(position == .left ? Constants.Strings.left : Constants.Strings.right)
                            .font(Constants.Fonts.button)
                            .foregroundColor(selection == position ? settings.backgroundColor : settings.textColor)
                            .frame(maxWidth: .infinity)
                            .frame(height: 60)
                            .background(selection == position ? settings.buttonColor : settings.buttonBackgroundColor)
                            .cornerRadius(Constants.Dimensions.cornerRadius)
                    }
                    .accessibilityLabel("\(position == .left ? "Left" : "Right") hand mode")
                    .accessibilityAddTraits(selection == position ? .isSelected : [])
                }
            }
        }
    }
}

struct SettingsToggle_Previews: PreviewProvider {
    static var previews: some View {
        VStack(spacing: 20) {
            SettingsToggle(title: "Light on at start", isOn: .constant(false))
            SettingsToggle(title: "High contrast mode", isOn: .constant(true))
            SettingsSegmentedControl(title: "Button Position", selection: .constant(.right))
        }
        .padding()
        .environmentObject(AppSettings())
    }
}
